game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java

```diff
@@ -668,92 +669,71 @@ private void removeAirNoLongerInTerritory() {
-          CollectionUtils.getMatches(this.attackingUnits, Matches.unitIsOwnedBy(current));
-      final String verb = current.equals(attacker) ? "attack" : "loiter and taunt";
-      transcriptText
-          .append(current.getName())
-          .append(" ")
-          .append(verb)
-          .append(
-              attackingUnits.isEmpty()
-                  ? ""
-                  : " with " + MyFormatter.unitsToTextNoOwner(attackingUnits))
-          .append(delim);
-      allAttackingUnits.addAll(attackingUnits);
-      // If any attacking transports are in the battle, set their status to later restrict
-      // load/unload
-      if (current.equals(attacker)) {
-        final CompositeChange change = new CompositeChange();
-        final Collection<Unit> transports =
-            CollectionUtils.getMatches(attackingUnits, Matches.unitCanTransport());
-        for (final Unit unit : transports) {
-          change.add(ChangeFactory.unitPropertyChange(unit, true, Unit.WAS_IN_COMBAT));
-        }
-        bridge.addChange(change);
+    // If any attacking transports are in the battle, set their status to later restrict
+    // load/unload
+    final Collection<Unit> transports =
+        CollectionUtils.getMatches(
+            attackingUnits, Matches.unitCanTransport().and(Matches.unitIsOwnedBy(attacker)));
+    if (!transports.isEmpty()) {
+      final CompositeChange change = new CompositeChange();
+      for (final Unit unit : transports) {
+        change.add(ChangeFactory.unitPropertyChange(unit, true, Unit.WAS_IN_COMBAT));
       }
+      bridge.addChange(change);
     }
-    // write attacking units to history
-    if (!attackingUnits.isEmpty()) {
-      bridge.getHistoryWriter().addChildToEvent(transcriptText.toString(), allAttackingUnits);
+  }
+
+  private void writeUnitsToHistory(final IDelegateBridge bridge) {
+    if (headless) {
+      return;
     }
-    // find all defending players (unsorted)
-    final Collection<GamePlayer> defenders = new ArrayList<>();
-    for (final GamePlayer current : playerSet) {
-      if (gameData.getRelationshipTracker().isAllied(defender, current)
-          || current.equals(defender)) {
-        defenders.add(current);
+    final Set<GamePlayer> playersWithUnits = battleSite.getUnitCollection().getPlayersWithUnits();
+
+    final Collection<GamePlayer> attackers = findAllies(playersWithUnits, attacker);
+    addPlayerCombatHistoryText(attackers, attackingUnits, true, bridge.getHistoryWriter());
+    final Collection<GamePlayer> defenders = findAllies(playersWithUnits, defender);
+    addPlayerCombatHistoryText(defenders, defendingUnits, false, bridge.getHistoryWriter());
+  }
+
+  private static Collection<GamePlayer> findAllies(
+      final Collection<GamePlayer> candidatePlayers, final GamePlayer player) {
+    final var relationshipTracker = player.getData().getRelationshipTracker();
+    final Collection<GamePlayer> allies = new ArrayList<>();
```

Comment:
FWIW, this block could be re-written as follows:
```
return candidatePlayers.stream()
    .filter(current -> current.equals(player) || relationshipTracker.isAllied(player, current)
    .collect(Collectors.toList());
```

I somewhat feel like relationship tracker probably should own this logic though. EG:
```
relationshipTracker.getAllies(player, candidatePlayers);
```

Reply:
There's certainly more places where we can simplify or 'fix' responsibilities and change logic ownership, but I recognize we are mostly moving code here.